lib/beaker/hypervisor/openstack.rb

```
@@ -250,83 +253,114 @@ def get_floating_ip
-
-        #set metadata
-        vm.metadata.update({:jenkins_build_url => @options[:jenkins_build_url].to_s,
-                            :department        => @options[:department].to_s,
-                            :project           => @options[:project].to_s })
-        @vms << vm
-
-        # Wait for the host to accept ssh logins
-        host.wait_for_port(22)
-
-        #enable root if user is not root
-        enable_root(host)
-
-        provision_storage(host, vm) if @options[:openstack_volume_support]
-        @logger.notify "OpenStack Volume Support Disabled, can't provision volumes" if not @options[:openstack_volume_support]
+        sleep SLEEPWAIT
+        try += 1
       end
-
+    
+      if @options[:openstack_floating_ip]
+        # Associate a public IP to the VM
+        ip.server = vm
+        host[:ip] = ip.ip
+      else
+        # Get the first address of the VM that was just created just like in the
+        # OpenStack UI
+        host[:ip] = vm.addresses.first[1][0]["addr"]
+      end
+    
+      @logger.debug "OpenStack host #{host.name} (#{host[:vmhostname]}) assigned ip: #{host[:ip]}"
+    
+      # Set metadata
+      vm.metadata.update({:jenkins_build_url => @options[:jenkins_build_url].to_s,
+                          :department        => @options[:department].to_s,
+                          :project           => @options[:project].to_s })
+      @vms << vm
+    
+      # Wait for the host to accept SSH logins
+      host.wait_for_port(22)
+    
+      # Enable root if the user is not root
+      enable_root(host)
+    
+      provision_storage(host, vm) if @options[:openstack_volume_support]
+      @logger.notify "OpenStack Volume Support Disabled, can't provision volumes" if not @options[:openstack_volume_support]
+    rescue => e
+      # Handle exceptions in the thread
+      puts "Thread #{host} failed with error: #{e.message}"
+    
       hack_etc_hosts @hosts, @options
```

This seems to also silently ignore exceptions and continue execution which can lead to unexpected errors later on.

Reading the diff is not easy, I feel like `hack_etc_hosts` happened on success and now happen when an exception is raised.  Weird.

---

yes, there were only 3 new functions and the previous instance create one moved down, but the diff looks awful